tests/integration/test_tls.py

```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/env python3
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import asyncio
+import logging
+from pathlib import Path
+
+import pytest
+import yaml
+from pytest_operator.plugin import OpsTest
+
+from tests.integration.helpers import check_properties, ping_servers
+
+logger = logging.getLogger(__name__)
+
+METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
+APP_NAME = METADATA["name"]
+
+
+@pytest.mark.abort_on_fail
+async def test_deploy_ssl_quorum(ops_test: OpsTest):
+    charm = await ops_test.build_charm(".")
+    await asyncio.gather(
+        ops_test.model.deploy(charm, application_name=APP_NAME, num_units=3),
+        ops_test.model.deploy(
+            "tls-certificates-operator",
+            application_name="tls-certificates-operator",
+            channel="edge",
+            num_units=1,
+            config={"generate-self-signed-certificates": "true", "ca-common-name": "zookeeper"},
+        ),
+    )
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME, "tls-certificates-operator"], status="active", timeout=1000
+    )
+    assert ops_test.model.applications[APP_NAME].status == "active"
+    assert ops_test.model.applications["tls-certificates-operator"].status == "active"
+    await ops_test.model.add_relation(APP_NAME, "tls-certificates-operator")
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME, "tls-certificates-operator"], status="active", timeout=1000
+    )
+    assert ops_test.model.applications[APP_NAME].status == "active"
+    assert ops_test.model.applications["tls-certificates-operator"].status == "active"
+
+    assert ping_servers(ops_test)
+
+    for unit in ops_test.model.applications[APP_NAME].units:
+        assert "sslQuorum=true" in check_properties(
+            model_full_name=ops_test.model_full_name, unit=unit.name
+        )
+
+
+@pytest.mark.abort_on_fail
+async def test_remove_tls_provider(ops_test: OpsTest):
+    await ops_test.model.remove_application("tls-certificates-operator", block_until_done=True)
+    await ops_test.model.wait_for_idle(apps=[APP_NAME])
+    assert ops_test.model.applications[APP_NAME].status == "active"
+
+    assert ping_servers(ops_test)
+
+    for unit in ops_test.model.applications[APP_NAME].units:
+        assert "sslQuorum=true" not in check_properties(
+            model_full_name=ops_test.model_full_name, unit=unit.name
+        )
+
+
+@pytest.mark.abort_on_fail
+async def test_add_tls_provider(ops_test: OpsTest):
```

Comment:
This tests seems to check the exact same thing the `test_deploy_ssl_quorum` checks. Is there a reason for that? I imagine that it is because of the next test (test_scale_up_tls), that needs TLS enabled. But in that case, is it possible to run `test_scale_up_tls` before `test_remove_tls_provider` to avoid this duplicate code?

Reply:
It's not clear from the name, but I was testing that I could remove, THEN add the provider. I had some bugs before where things would fail because, for example, certs/keys were still on the unit, so this was more of a defensive tests against that. I will rename the test though to make that clearer.

Reply:
Nice! It makes total sense to have a check like that. A rename like you commented or also a comment in the test should help with that understanding.